packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,163 @@ impl<'m> RunnerImpl<'m> {
+                };
+                builder = builder.add_buffer(Buffer::from_slice_ref(&offsets));
+
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
+                    inner_field,
                     None,
-                )?])
+                )?);
             }
-            DataType::FixedSizeList(field, multiplier) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+            DataType::FixedSizeList(inner_field, size) => {
+                // FixedSizeListList is only stored by child data, as offsets are not required
+                // because the size is known.
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
-                    Some(data.len * multiplier as usize),
-                )?])
+                    inner_field,
+                    Some(target_len * *size as usize),
+                )?);
             }
-            DataType::Struct(fields) => {
-                let mut v = Vec::new();
-                for (i, field) in fields.iter().enumerate() {
+            DataType::Struct(inner_fields) => {
+                // Structs are only defined by child data
+
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                for (i, inner_field) in inner_fields.iter().enumerate() {
                     let child = child_data.get(i as u32)?;
-                    v.push(self.array_data_from_js(
+                    builder = builder.add_child_data(self.array_data_from_js(
                         mv8,
                         &child,
-                        field.data_type(),
-                        Some(data.len),
+                        inner_field,
+                        Some(target_len),
                     )?);
                 }
-                Ok(v)
             }
-            t => {
-                if n_children == 0 {
-                    Ok(vec![])
-                } else {
-                    Err(Error::FlushType(t))
-                }
+            DataType::FixedSizeBinary(size) => {
+                // FixedSizeBinary is only stored as a buffer (u8), offsets are not required because
+                // the size is known
+
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u8`, and
+                //   `data.len * *size as usize` corresponds for the `size` of *each* value.
+                let values = unsafe {
+                    slice::from_raw_parts(data.buffer_ptrs[0], data.len * *size as usize)
```

Comment:
```suggestion
                //   `*size as usize * data.len` corresponds for the `size` of *each* value.
                let values = unsafe {
                    slice::from_raw_parts(data.buffer_ptrs[0], *size as usize * data.len)
```

Slightly more readable